tests/core/full_node/stores/test_coin_store.py

```diff
@@ -493,6 +495,186 @@ async def test_get_coin_states(db_version: int) -> None:
         assert len(await coin_store.get_coin_states_by_ids(True, coins, uint32(0), max_items=10000)) == 600
 
 
+@pytest.mark.anyio
+async def test_coin_state_batches(db_version: int) -> None:
+    async with DBConnection(db_version) as db_wrapper:
+        # Generate coin records, puzzle hashes, and hints.
+        coin_records: List[CoinRecord] = []
+        puzzle_hashes: List[bytes32] = []
+        hints: List[Tuple[bytes32, bytes]] = []
+        count = 50000
+
+        for i in range(count):
+            is_spent = i % 2 == 0
+            is_hinted = i % 7 == 0
+            created_height = uint32(i % 10)
+            spent_height = uint32(created_height + 100)
+
+            puzzle_hash = std_hash(i.to_bytes(4, byteorder="big"))
+
+            coin = Coin(
+                std_hash(b"Parent Coin Id " + i.to_bytes(4, byteorder="big")),
+                puzzle_hash,
+                uint64(i),
+            )
+
+            if is_hinted:
+                hint = std_hash(b"Hinted " + puzzle_hash)
+                hints.append((coin.name(), hint))
+                puzzle_hashes.append(hint)
+            else:
+                puzzle_hashes.append(puzzle_hash)
+
+            coin_records.append(
+                CoinRecord(
+                    coin=coin,
+                    confirmed_block_index=created_height,
+                    spent_block_index=spent_height if is_spent else uint32(0),
+                    coinbase=False,
+                    timestamp=uint64(0),
+                )
+            )
+
+        # Initialize coin and hint stores.
+        coin_store = await CoinStore.create(db_wrapper)
+        hint_store = await HintStore.create(db_wrapper)
+
+        await coin_store._add_coin_records(coin_records)
+        await hint_store.add_hints(hints)
+
+        # Helper for syncing all of the coin states.
+        async def test_states(
+            expected_records: List[CoinRecord], *, include_spent: bool, include_unspent: bool, include_hinted: bool
+        ) -> None:
+            height = uint32(0)
+            all_coin_states: List[CoinState] = []
+            is_finished = False
+            remaining_phs = puzzle_hashes.copy()
+
+            while not is_finished:
+                (coin_states, height, is_finished) = await coin_store.batch_coin_states_by_puzzle_hashes(
+                    remaining_phs[:15000],
+                    min_height=height,
+                    include_spent=include_spent,
+                    include_unspent=include_unspent,
+                    include_hinted=include_hinted,
+                )
+                all_coin_states += coin_states
+
+                if is_finished:
+                    remaining_phs = remaining_phs[15000:]
```

Comment:
what happens if the number of puzzle hashes exceed the limit? Is there an assert?
Now it looks like the caller can just assume that all puzzle hashes that were requested will be returned.

Reply:
Yes, there's an assertion to prevent this. It's up to the caller of the batch coin states method to know this based on the doc string, or it will raise an error.

Reply:
The wallet protocol will handle this gracefully though